flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/generated/HashFunction.java

```diff
@@ -18,12 +18,23 @@
 
 package org.apache.flink.table.runtime.generated;
 
+import org.apache.flink.table.data.ArrayData;
+import org.apache.flink.table.data.MapData;
 import org.apache.flink.table.data.RowData;
 
 /**
- * Interface for code generated hash code of {@link RowData}, which will select some fields to hash.
+ * Interface for code generated hash code of {@link RowData} which will select some fields to hash
+ * or {@link ArrayData} or {@link MapData}.
+ *
+ * <p>Due to Janino's support for generic type is not very friendly, so here can't introduce generic
+ * type for {@link HashFunction}, please see https://github.com/janino-compiler/janino/issues/109
+ * for details.
```

Comment:
You can define generic types in the interface. The Janino limitation requires generating the method parameters using `Object` types instead of generic types. However, that's not a problem. You can manually cast the input parameter to `RowData`/`ArrayData`/`MapData` type in the first line of the method. See `org.apache.flink.table.planner.codegen.FunctionCodeGenerator#generateFunction`

I don't like overloading many methods, but just one method is valid.